samcli/commands/publish/command.py

```diff
@@ -0,0 +1,138 @@
+@pass_context
+def cli(ctx, template):
+    # All logic must be implemented in the ``do_cli`` method. This helps with easy unit testing
+
+    do_cli(ctx, template)  # pragma: no cover
+
+
+def do_cli(ctx, template):
+    """Publish the application based on command line inputs."""
+    if not os.path.exists(template):
+        click.secho("Publish Failed", fg='red')
+        raise SamTemplateNotFoundException("Template at {} is not found".format(template))
+
+    with click.open_file(template, 'r') as template_file:
+        try:
+            output = publish_application(template_file.read())
+            click.secho("Publish Succeeded", fg="green")
+            click.secho(_gen_success_message(output), fg="yellow")
+        except ServerlessRepoError as ex:
+            click.secho("Publish Failed", fg='red')
+            raise UserException(str(ex))
+        except ClientError as ex:
+            click.secho("Publish Failed", fg='red')
+            raise _wrap_s3_uri_exception(ex)
+
+        application_id = output['application_id']
+        _print_console_link(ctx.region, application_id)
+
+
+def _gen_success_message(publish_output):
+    """
+    Generate detailed success message for published applications.
+
+    Parameters
+    ----------
+    publish_output : dict
+        Output from serverlessrepo publish_application
+
+    Returns
+    -------
+    str
+        Detailed success message
+    """
+    application_id = publish_output['application_id']
+    details = json.dumps(publish_output['details'], indent=2)
+
+    if CREATE_APPLICATION in publish_output['actions']:
+        return "Created new application with the following metadata:\n{}".format(details)
+
+    return 'The following metadata of application "{}" has been updated:\n{}'.format(application_id, details)
+
+
+def _print_console_link(region, application_id):
+    """
+    Print link for the application in AWS Serverless Application Repository console.
+
+    Parameters
+    ----------
+    region : str
+        AWS region name
+    application_id : str
+        The Amazon Resource Name (ARN) of the application
+
+    """
+    if not region:
+        region = boto3.Session().region_name
+    url = "https://console.aws.amazon.com/serverlessrepo/home?region={}#/published-applications/{}"
+    console_link = url.format(region, application_id.replace('/', '~'))
+    msg = "Click the link below to view your application in AWS console:\n{}".format(console_link)
+    click.secho(msg, fg="yellow")
+
+
+def _wrap_s3_uri_exception(ex):
+    """
+    Wrap invalid S3 URI exception with a better error message.
+
+    Parameters
+    ----------
+    ex : ClientError
+        boto3 exception
+
+    Returns
+    -------
+    Exception
+        UserException if found invalid S3 URI or ClientError
+    """
+    error_code = ex.response['Error']['Code']
+    message = ex.response['Error']['Message']
+
+    if error_code == 'BadRequestException' and "Invalid S3 URI" in message:
```

Comment:
Is the only way of knowing this through the message? Seems kind of fragile

Reply:
Yeah, this is the only way.

Reply:
:(